python/src/opendp/mod.py

```diff
@@ -1377,6 +1361,37 @@ def _exception_predicate(v):
     return signed_band_search(center, at_center, sign)
 
 
+def _infer_type(predicate: Callable[[float], bool]) -> Union[Type[float], Type[int]]:
+    def _is_type_error(e):
+        is_match_error = isinstance(e, OpenDPException) and "No match for concrete type" in (e.message or "")
+        is_type_error = isinstance(e, TypeError)
+        return is_match_error or is_type_error
+
+    try:
+        predicate(0.)
+        # predicate succeeded with a float, assume it wants floats
+        return float
+    except Exception as e:
+        if not _is_type_error(e):
+            # it didn't reject the type, but did fail with a different error
+            return float
+    
+    # Try again with the more forgiving type, an int.
+    try:
+        predicate(0)
+        # predicate is happy with an int, assume it wants ints
+        return int
+    except Exception as e:
+        if not _is_type_error(e):
+            # it didn't reject the type, but did fail with a different error
+            return int
+
+        # enrich the error message if in Python 3.11+.
+        if hasattr(e, "add_note"):
+            e.add_note("Unable to infer type `T`; pass the type `T` or bounds. This exception is raised when the predicate is evaluated at 0")
+        raise
```

Comment:
Would [exception chaining](https://docs.python.org/3/tutorial/errors.html#exception-chaining) be better here? 

We have no idea what exception was thrown by `predicate`, but whatever it was doesn't matter at this point: We want to tell the user there's a `TypeError` or something similar.

Reply:
The exception thrown by the predicate is usually much more interesting! I typically don't care about the trace leading to the binary search.